cmd/tier-journal.go

```diff
@@ -0,0 +1,278 @@
+)
+
+//go:generate msgp -file $GOFILE -unexported
+//msgp:ignore tierJournal walkfn
+
+type tierJournal struct {
+	sync.RWMutex
+	diskPath string
+	file     *os.File // active journal file
+}
+
+type jentry struct {
+	ObjName  string `msg:"obj"`
+	TierName string `msg:"tier"`
+}
+
+const (
+	tierJournalVersion = 1
+	tierJournalHdrLen  = 2 // 2 bytes
+)
+
+func initTierDeletionJournal(done <-chan struct{}) (*tierJournal, error) {
+	diskPath, ok := globalEndpoints.FirstLocalDiskPath()
+	if !ok {
+		return nil, errors.New("failed to find first local disk path")
+	}
+
+	j := &tierJournal{
+		diskPath: diskPath,
+	}
+
+	if err := os.MkdirAll(filepath.Dir(j.JournalPath()), os.FileMode(0770)); err != nil {
+		return nil, err
+	}
+
+	err := j.Open()
+	if err != nil {
+		return nil, err
+	}
+
+	go j.deletePending(done)
+	return j, nil
+}
+
+// rotate rotates the journal. If a read-only journal already exists it does
+// nothing. Otherwise renames the active journal to a read-only journal and
+// opens a new active journal.
+func (j *tierJournal) rotate() error {
+	// Do nothing if a read-only journal file already exists.
+	if _, err := os.Stat(j.ReadOnlyPath()); err == nil {
+		return nil
+	}
+	// Close the active journal if present.
+	j.Close()
+	// Open a new active journal for subsequent journalling.
+	return j.Open()
+}
+
+type walkFn func(objName, tierName string) error
+
+func (j *tierJournal) ReadOnlyPath() string {
+	return filepath.Join(j.diskPath, minioMetaBucket, "ilm", "deletion-journal.ro.bin")
+}
+
+func (j *tierJournal) JournalPath() string {
+	return filepath.Join(j.diskPath, minioMetaBucket, "ilm", "deletion-journal.bin")
+}
+
+func (j *tierJournal) WalkEntries(fn walkFn) {
+	err := j.rotate()
+	if err != nil {
+		logger.LogIf(context.Background(), fmt.Errorf("tier-journal: failed to rotate pending deletes journal %s", err))
+		return
+	}
+
+	ro, err := j.OpenRO()
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		return // No read-only journal to process; nothing to do.
+	case err != nil:
+		logger.LogIf(context.Background(), fmt.Errorf("tier-journal: failed open read-only journal for processing %s", err))
+		return
+	}
+	defer ro.Close()
+	mr := msgp.NewReader(ro)
+	done := false
+	for {
+		var entry jentry
+		err := entry.DecodeMsg(mr)
+		if errors.Is(err, io.EOF) {
+			done = true
+			break
+		}
+		if err != nil {
+			logger.LogIf(context.Background(), fmt.Errorf("tier-journal: failed to decode journal entry %s", err))
+			break
+		}
+		err = fn(entry.ObjName, entry.TierName)
+		if err != nil && !isErrObjectNotFound(err) {
+			logger.LogIf(context.Background(), fmt.Errorf("tier-journal: failed to delete transitioned object %s from %s due to %s", entry.ObjName, entry.TierName, err))
+			j.AddEntry(entry)
+		}
+	}
+	if done {
+		os.Remove(j.ReadOnlyPath())
+	}
+}
+
+func deleteObjectFromRemoteTier(objName, tierName string) error {
+	w, err := globalTierConfigMgr.getDriver(tierName)
+	if err != nil {
+		return err
+	}
```

Comment:
if tier name is not found, seems to me we should just drop the entry rather than recirculating it in the journal forever. thoughts?